### PR TITLE
[OP] Unsorted segment sum1D

### PR DESCRIPTION
**Pull request by kmagiers**

### CODE REVIEW

**Comment by marvin-Yu** [comment]
Hi @kmagiers, I`m planning to backport the parallel optimize feature at #59, could you please make sure if it will affect your work here?

**Comment by shanzhou2186** [comment]
@kmagiers  Could you help to supply the performance result of this optimization, please? If you have the data based on operator   and model posted in jira or something else, please share the link. Thanks.

**Comment by kmagiers** [comment]
> 

performance data for this approach was provided in this: https://jira.devtools.intel.com/browse/WC1-1160 JIRA ticket

**Comment by changqi1** [comment]
@kmagiers See you to use
template <typename Index>
class UnsortedSegmentReductionOp<float, Index, USS_FP32<Index>> {};
    to do 
template <typename T, typename Index, typename DeviceReductionFunctor>
class UnsortedSegmentReductionOp {};
    specialization for data_is_1D.

This is a good point to distinguish your optimization. But it couldn't have been better maintained. Meanwhile, we need to update both two operators if we have some updates. So would you add your optimization in default op:
template <typename T, typename Index, typename DeviceReductionFunctor>
class UnsortedSegmentReductionOp {};

**Comment by changqi1** [comment]
@kmagiers  Do you test multi-thread perf? like this one: https://github.com/changqi1/DeepRec/blob/main/tensorflow/core/kernels/mkl_matmul_op_test.cc#L267-L293

**Comment by changqi1** [comment]
@marvin-Yu Do you think the tensorflow/core/util/test_util.h file is necessary only for this op's UT? Would we reuse tensorflow existing API to implement it? Please give some guides to @kmagiers .

**Comment by kmagiers** [comment]
> @marvin-Yu Do you think the tensorflow/core/util/test_util.h file is necessary only for this op's UT? Would we reuse tensorflow existing API to implement it? Please give some guides to @kmagiers .

This file have only versalite code that can (and should) be reused in all future unit tests, also most already existing unit tests created by our team can be refactored to use this file instead of copy pasting code (RunAndFetch and tensor from std::vector creation for example)

**Comment by kmagiers** [comment]
> @kmagiers Do you test multi-thread perf? like this one: https://github.com/changqi1/DeepRec/blob/main/tensorflow/core/kernels/mkl_matmul_op_test.cc#L267-L293



> @kmagiers Do you test multi-thread perf? like this one: https://github.com/changqi1/DeepRec/blob/main/tensorflow/core/kernels/mkl_matmul_op_test.cc#L267-L293

I did not really touch benchmark code, just added new test cases and rerun existing benchmark, it seems that it does not test intra paralleling (If I am not mistaken this op at this stage does not support parallel run so no point in this I guess)

**Comment by changqi1** [comment]
> > @marvin-Yu Do you think the tensorflow/core/util/test_util.h file is necessary only for this op's UT? Would we reuse tensorflow existing API to implement it? Please give some guides to @kmagiers .
> 
> This file have only versalite code that can (and should) be reused in all future unit tests, also most already existing unit tests created by our team can be refactored to use this file instead of copy pasting code (RunAndFetch and tensor from std::vector creation for example)

Yes, I know your efforts. Do you see or refer this UT code: https://github.com/changqi1/DeepRec/blob/main/tensorflow/core/kernels/segment_reduction_ali_ops_test.cc

**Comment by kmagiers** [comment]
> > > @marvin-Yu Do you think the tensorflow/core/util/test_util.h file is necessary only for this op's UT? Would we reuse tensorflow existing API to implement it? Please give some guides to @kmagiers .
> > 
> > 
> > This file have only versalite code that can (and should) be reused in all future unit tests, also most already existing unit tests created by our team can be refactored to use this file instead of copy pasting code (RunAndFetch and tensor from std::vector creation for example)
> 
> Yes, I know your efforts. Do you see or refer this UT code: https://github.com/changqi1/DeepRec/blob/main/tensorflow/core/kernels/segment_reduction_ali_ops_test.cc

The UT that you gave have nothing in common with this one, even though the tests have "segment" in name those are sparse format operations (https://www.tensorflow.org/guide/sparse_tensor)

**Comment by kmagiers** [comment]
> @kmagiers See you to use template class UnsortedSegmentReductionOp<float, Index, USS_FP32> {}; to do template <typename T, typename Index, typename DeviceReductionFunctor> class UnsortedSegmentReductionOp {}; specialization for data_is_1D.
> 
> This is a good point to distinguish your optimization. But it couldn't have been better maintained. Meanwhile, we need to update both two operators if we have some updates. So would you add your optimization in default op: template <typename T, typename Index, typename DeviceReductionFunctor> class UnsortedSegmentReductionOp {};

Done, confirmed that no benchamrks/tests fail after changes, performance is similar to original solution

File: tensorflow/core/kernels/segment_reduction_ops.cc [hunk 1]
```diff
@@ -510,6 +535,15 @@ class UnsortedSegmentReductionOp : public OpKernel {
   DeviceReductionFunctor reduction_functor_;
 };
 
+template <typename T>
+using USS_FP32 = functor::UnsortedSegmentFunctor<
```
Comment (changqi1) [inline tensorflow/core/kernels/segment_reduction_ops.cc:539]: where will the 'USS_FP32'  be used?

**Comment by changqi1** [comment]
> > @kmagiers See you to use template class UnsortedSegmentReductionOp<float, Index, USS_FP32> {}; to do template <typename T, typename Index, typename DeviceReductionFunctor> class UnsortedSegmentReductionOp {}; specialization for data_is_1D.
> > This is a good point to distinguish your optimization. But it couldn't have been better maintained. Meanwhile, we need to update both two operators if we have some updates. So would you add your optimization in default op: template <typename T, typename Index, typename DeviceReductionFunctor> class UnsortedSegmentReductionOp {};
> 
> Done, confirmed that no benchamrks/tests fail after changes, performance is similar to original solution

@kmagiers  Please push your PR to alibaba/DeepRec main branch.

**Comment by changqi1** [comment]
submitted in https://github.com/alibaba/DeepRec/pull/240

**Pull request closed without merging**